subsys: net: lib: Add AWS IoT library

simensrostad opened the pull request:
Add AWS IoT library

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## Gitlint issues

1: UC3 Title does not follow [subsystem]: [subject] (and should not start with literal subsys:): "subsys: net: lib: Add bifravst_cloud"



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

rlubos commented on subsys/net/lib/bifravst_cloud/src/bifravst_cloud.c:
```diff
@@ -0,0 +1,546 @@
+	char imei_buf[AWS_CLOUD_CLIENT_ID_LEN + 1];
+
+	err = at_cmd_write("AT+CGSN", imei_buf,
+		(AWS_CLOUD_CLIENT_ID_LEN + 5), NULL);
```
Why is the buffer size provided to `at_cmd_write` larger than the actual `imei_buf` size? This can cause stack corruption.

The cleanest approach would be to provide `sizeof(imei_buf)` to the `at_cmd_write`. If you need a larger buffer, just make it larger.

rlubos commented on subsys/net/lib/bifravst_cloud/src/bifravst_cloud.c:
```diff
@@ -0,0 +1,546 @@
+		return err;
+	}
+
+	err = snprintf(id, sizeof(client_id_buf), "%s", imei_buf);
```
It doesn't make much sense to mix statically allocated buffer size (`sizeof(client_id_buf)`) with theoretically generic buffer provided through a pointer.

I'd either provide additional `len` parameter to the function, or gave up on parameters at all and operate directly on `client_id_buf`.

rlubos replied:
And if the content of `imei_buf` is a string anyway, couldn't the `at_cmd_write` write directly into `client_id_buf`? I don't see a need for extra copy here, unless I miss something?

simensrostad replied:
You are right. Writing directly to ```client_id_buf``` makes more sense and looks cleaner.

rlubos commented on subsys/net/lib/bifravst_cloud/src/bifravst_cloud.c:
```diff
@@ -0,0 +1,546 @@
+	const struct mqtt_subscription_list subscription_list = {
+		.list = (struct mqtt_topic *)&cc_rx_list,
+		.list_count = ARRAY_SIZE(cc_rx_list),
+		.message_id = SUBSCRIBE_ID
```
Why this specific `message_id` is used instead of a random number?

rlubos commented on subsys/net/lib/bifravst_cloud/src/bifravst_cloud.c:
```diff
@@ -0,0 +1,546 @@
+static int broker_init(void)
+{
+	int err;
+
```
Extra empty line.

rlubos commented on subsys/net/lib/bifravst_cloud/src/bifravst_cloud.c:
```diff
@@ -0,0 +1,546 @@
+{
+	int err;
+
+	struct sockaddr_in *broker =
```
Could we change this variable name to `broker4`? Shadowing the global variable name makes it slightly confusing.

rlubos commented on subsys/net/lib/bifravst_cloud/src/bifravst_cloud.c:
```diff
@@ -0,0 +1,546 @@
+
+			inet_ntop(AF_INET6, &broker6->sin6_addr.s6_addr,
+				ipv6_addr,
+				sizeof(ipv6_addr));
```
This would fit in the line above.

rlubos commented on subsys/net/lib/bifravst_cloud/src/bifravst_cloud.c:
```diff
@@ -0,0 +1,546 @@
+	client->tx_buf_size		= sizeof(tx_buffer);
+	client->transport.type		= MQTT_TRANSPORT_SECURE;
+
+	static sec_tag_t sec_tag_list[] = { CONFIG_BIFRAVST_CLOUD_SEC_TAG };
```
What credential does this cloud use, how is it provisioned? Will a sample follow, showing the use of this cloud?

simensrostad replied:
I have contacted you on Microsoft teams regarding provisioning, documentation and sample

simensrostad commented on subsys/net/lib/bifravst_cloud/Kconfig:
```diff
@@ -0,0 +1,54 @@
+menuconfig BIFRAVST_CLOUD
+	bool "BIFRAVST Cloud library"
+	select CLOUD_API
+	select CJSON_LIB
```
It should not be. It is residue from an earlier implementation.

jtguggedal commented:
@rlubos I think you'll need to revisit, there's been significant changes. It's more of a generic AWS IoT backend now.

rlubos commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+	}
+
+#if defined(CONFIG_AWS_IOT_GET_ACCEPTED_TOPIC_SUBSCRIBE)
+	err = snprintf(get_accepted_topic, sizeof(get_accepted_topic), GET_ACCEPTED_TOPIC,
```
This and a few lines below are quite lengthy, remember we have 80 characters/line limit.

rlubos commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+	return err;
+}
+
+static int aws_iot_ep_topics_set(const struct cloud_backend *const backend,
```
How is this function used? I hope it's not called more than once, otherwise, we'll have a memory leak. Possibly, you could `free` the `app_topic_data.list` if it is not NULL, before allocating a new one.

I guess we should also have an `uninit` function that would do the cleanup.

simensrostad replied:
This function is called by the application during initialization to pass in a list of subscription topics which will be subscribed to every time the AWS IoT library gets a CONNACK MQTT event. Its intention is to be called only once. However the Cloud API does have an ```uninit`` function which I can attach to a new function in the AWS IoT lib which frees the ```app_topic_data.list```. This will make it possible to change the application specific subscribed topics at run-time. Do you think that the already existing ```uninit``` should be used or shall I make an uninit function spesifically for freeing the ```app_topic_data.list```?

rlubos replied:
I was thinking about re-using the existing `uninit` API - it should clean up any mess that cloud engine made during its operation. I can imagine a situation when we init/unint the cloud engine several times during a single runtime cycle, and in that case, it's important to have a proper cleanup.

And as for calling malloc within `CONNACK MQTT` event handler - are you sure it will happen only once during application runtime? The cloud could disconnect over time for various reasons, I'd imagine to re-connect in such case instead of rebooting entire application. And in such case we'll get another CONNACK event .

simensrostad replied:
There are two functions that "manage" the subscription list. ```aws_iot_topics_subscribe``` allocates a list of application spesific topics in the initialization phase of the app(linked through the cloud api) and one internal static function ```topics_subscribe``` which is called every time the aws iot library gets a connack. ```topics_subscribe``` does not allocate memory, it just subscribes to topics depending on the content of aws shadow topics(Kconfigurable) and application spesific topic list (which ```aws_iot_topics_subscribe``` allocates memory for). The memory leak will occur only if the application calls ```cloud_topics_subscribe``` more than once without freeing the previous allocated list.

jtguggedal replied:
I'm late to the party here, but would be nice to avoid the heap entirely. I know you've mem slabs API a bit, @rlubos . Is that a candidate here?

rlubos replied:
If it's desired to avoid heap usage here, I'd consider static allocation in this particular use-case. It only makes sense to have a single `app_topic_data.list` available at a time, and the only variable here is number of topics registered.

We could introduce a Kconfig option to specify the maximum number of topics allowed, and simply pre-allocate the list in static memory. While a mem slab could also be considered here, I think it has no added value for this use-case - it would only complicate the code a bit, and we would need to apply some limits on it anyway.

jtguggedal replied:
Agreed, (Kconfigurable sized) static allocation is reasonable.

simensrostad replied:
That sounds like a reasonable alternative. Ill make the appropriate changes.

rlubos commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+
+		for (size_t i = 0; i < app_sub_list.list_count; i++) {
+			LOG_DBG("Subscribing to application spesific topic: %s",
+			log_strdup(app_sub_list.list[i].topic.utf8));
```
Alignment.

rlubos commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+		LOG_ERR("aws_fota_mqtt_evt_handler, error: %d", err);
+		cloud_evt.type = CLOUD_EVT_ERROR;
+		cloud_notify_event(aws_iot_backend, &cloud_evt,
+				   config->user_data);
```
Shouldn't we quit on error as well? I guess that if `aws_fota` threw an error, it has handled the event.

simensrostad replied:
This is now changed according to the ```aws_fota``` example where the device disconnects from MQTT upon a ```FOTA_ERROR``` event.

simensrostad commented on include/net/cloud.h:
```diff
@@ -281,6 +290,24 @@ static inline int cloud_input(const struct cloud_backend *const backend)
+ * @param endpoint Pointer to list of endpoint topics.
+ * @param topics_count Number of endpoint topics in list
+ */
+static inline int cloud_ep_topics_set(const struct cloud_backend *const backend,
```
Yes, it makes sense to keep away technology specific concepts in a generic API. In the spirit of generality I will change it to your proposal. I was under the assumption that the Cloud API would only be used with pub/sub cloud architectures, where "topics" are common.

simensrostad replied:
```cloud_ep_subscriptions_set``` is a little bit lengthy forcing all the input argument to have their own line. Even cloud_subscriptions_set is a little bit long. How about ```cloud_sub_list_set```?

rlubos replied:
I'm fine with the shorter form, but let's keep this thread open for other proposals. Maybe someone will come up with some brilliant name for this API function.

jtguggedal replied:
I think my lacking vocabulary when naming this `endpoints` in the first place is coming back to confuse us now. Endpoint in the cloud API context was originally intended to cover topics in MQTT, resources in CoaP, URI in REST based APIs and so on. So `_endpoints_topics_` would thus be sort of redundant naming, or at least specific for MQTT.

If I understand the intention of this function correctly, it's a way to configure endpoints for incoming traffic from a cloud. It would be possible to add endpoints on runtime, ie the list does not have to be exhaustive at compile time, and the function can be called multiple times without overwritin existing endpoints. If that understanding is correct, perhaps something like `cloud_ep_subscriptions_add` is a fit? We would need anouther `_remove` / `_delete` to go along with it, though.

simensrostad replied:
Is it really necessary with a corresponding delete functionality if all this can be done with the existing ```uninit``` API? I agree in the sense that it looks cleaner if ```cloud_ep_subscriptions_add``` had its own ```_remove```.

jtguggedal replied:
What I had in mind for the `_delete` API was to, in MQTT context, unsubscribe from a topic to not waste data traffic on that topic if the device at some point doesn't need the data from the topic anymore

jtguggedal replied:
So, an operation on a single endpoint, not wiping the entire list

jtguggedal replied:
Perhaps not a super-important feature

jtguggedal commented on subsys/net/lib/aws_iot/Kconfig:
```diff
@@ -0,0 +1,75 @@
+	select CLOUD_API
+	select MQTT_LIB
+	select MQTT_LIB_TLS
+	select AT_CMD
```
I think we should try to avoid this dependency to `AT_CMD`

simensrostad replied:
In that case, is there any alternative way of getting the device IMEI for the shadow topics?

jtguggedal replied:
I think this should perhaps go into a new field in `struct cloud_backend_config`. Alternatively, the `user_data` pointer can be used for whater one wishes, for example a null-terminated string containing user ID. A bit hacky, perhaps, but doable.

jtguggedal commented on subsys/net/lib/aws_iot/Kconfig:
```diff
@@ -0,0 +1,75 @@
+
+config AWS_IOT_HOST_NAME
+	string "AWS IoT server hostname"
+	default "a34x44yyrk96tg-ats.iot.eu-central-1.amazonaws.com"
```
Not sure we should provide a default value here, perhaps better to fail compile time with a sensible warning than attempt connection to a random hostname

simensrostad replied:
Agreed

jtguggedal commented on subsys/net/lib/aws_iot/Kconfig:
```diff
@@ -0,0 +1,75 @@
+	default 2048
+	help
+	  Specifies maximum message size can be transmitted/received through
+	  MQTT (exluding MQTT PUBLISH payload).
```
Is the required buffer size that big if the PUBLISH payload is excluded? By the way, does PUBLISH in this context mean the payload received in an `MQTT_PUBLISH` event, or what's sent in a publish message?

simensrostad replied:
The payload buffer is the buffer populated during a ```PUBLISH``` event yes. So receiving.

simensrostad replied:
Im not sure I understand the first question there.

jtguggedal replied:
I was referring to the help text: `... (exluding MQTT PUBLISH payload)`. If the publish payload is not supposed to go into this buffer, then I assume it's intended for storing the MQTT header and for example payload in a `SUBSCRIBE` message. In that case, default of 2048 seems excessive.

jtguggedal commented on subsys/net/lib/aws_iot/Kconfig:
```diff
@@ -0,0 +1,75 @@
+
+config AWS_IOT_MQTT_PAYLOAD_BUFFER_LEN
+	int "Size of the buffer for MQTT PUBLISH payload."
+	default 2048
```
I think `AWS_IOT_MQTT_RX_TX_BUFFER_LEN + AWS_IOT_MQTT_PAYLOAD_BUFFER_LEN` cannot exceed the limitations in the modem of ~2.3k. That means, the Kconfigs can exceed it, but messages that big will never be received.

simensrostad replied:
Yes, that makes sense. Those sizes were set just to make sure the size limit in those buffers would cause problems during development. I will set some reasonable stock values here, 256 for example.

simensrostad replied:
Yes, that makes sense. Those sizes were set just to make sure the size limit in those buffers would cause problems during development. I will set some reasonable stock values here, 256 for example.

jtguggedal commented on subsys/net/lib/aws_iot/Kconfig:
```diff
@@ -0,0 +1,75 @@
+config AWS_IOT_IPV6
+	bool "Configure AWS IoT library to use IPv6 addressing. Otherwise IPv4 is used."
+
+config AWS_IOT_GET_ACCEPTED_TOPIC_SUBSCRIBE
```
Could we switch this around a bit, and use `AWS_IOT_TOPIC_SUB_...` for a bit of namespacing?

jtguggedal commented on subsys/net/lib/aws_iot/Kconfig:
```diff
@@ -0,0 +1,75 @@
+	bool "Configure AWS IoT library to use IPv6 addressing. Otherwise IPv4 is used."
+
+config AWS_IOT_GET_ACCEPTED_TOPIC_SUBSCRIBE
+	bool "Subscribe to get accepted shadow topic"
```
Let's do something like `$aws/things/<thing-name>/shadow/get/accepted` or similar to be explicit what it means

jtguggedal commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+#include <net/socket.h>
+#include <net/cloud.h>
+#include <net/cloud_backend.h>
+#include <at_cmd.h>
```
Wondering if we can have a Kconfig that includes this if the user wants to have IMEI as ID

jtguggedal commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+#endif
+
+#define AWS_TOPIC_PREFIX "$aws/things/"
+#define AWS_LEN (sizeof(AWS_TOPIC_PREFIX) - 1)
```
`AWS_TOPIC_PREFIX_LEN`

jtguggedal commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+#define AWS_TOPIC_PREFIX "$aws/things/"
+#define AWS_LEN (sizeof(AWS_TOPIC_PREFIX) - 1)
+
+#define AWS_CLIENT_ID_LEN 15
```
I think the user should be able to decide ID, so this should not be hard-coded

jtguggedal commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+#define DELETE_TOPIC AWS_TOPIC_PREFIX "%s/shadow/delete"
+#define DELETE_TOPIC_LEN (AWS_LEN + AWS_CLIENT_ID_LEN + 14)
+
+static char client_id_buf[AWS_CLIENT_ID_LEN + 5];
```
A few magic numbers here that might confuse the reader

jtguggedal commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+struct aws_iot_tx_data {
+	char *buf;
+	size_t len;
+	char *topic;
```
Is this always 0-terminated?

jtguggedal commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+	size_t list_count;
+};
+
+static struct app_topic_data app_topic_data = {.list_count = 0};
```
Nit: A little bit of air is always nice for these: `{ .list_count = 0 }`

jtguggedal commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+		};
+
+		for (size_t i = 0; i < aws_sub_list.list_count; i++) {
+			LOG_DBG("Subscribing to aws spesific topic: %s",
```
`AWS`

jtguggedal commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+		cloud_evt.data.msg.len = p->message.payload.len;
+
+		cloud_notify_event(aws_iot_backend, &cloud_evt,
+				config->user_data);
```
Alignment?

jtguggedal commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,679 @@
+	}
+
+	return data_publish(&client, tx_data.qos, tx_data.buf,
+				tx_data.len, tx_data.topic);
```
Alignment

simensrostad commented:
I have updated the library with the following  "major" changes:
 - Static allocation of the list which holds application specific topics
 - It is now possible use a static client id or a client id passed through the cloud API by the application, this option is kconfigurable. The chosen client id is used as mqtt client id and replaces <thing-name> in shadow topics.
 - Added functions for adding and removing endpoint subscriptions in generic cloud API.
 - Added sample which demonstrates the use of the ```aws_iot``` library,   ```aws_iot_simple``` sample

Both the  ```aws_iot_simple``` sample and  ```aws_iot``` library still needs documentation.

@rlubos @jtguggedal

rlubos commented on include/net/aws_iot.h:
```diff
@@ -0,0 +1,185 @@
+ * @param[in] event_handler Pointer to event handler to receive AWS IoT module
+ *                      events.
+ *
+ * @retval 0 If successful.
```
`@retval` should be used when you list return values (and each entry should start with it's own `@retval` marker). In case you provide a single description, you should use `@return` instead.

rlubos commented on include/net/aws_iot.h:
```diff
@@ -0,0 +1,185 @@
+/** @brief Send data to AWS IoT broker.
+ *
+ * @param[in] tx_data Pointer to struct containing data to be transmitted to
+ *                the AWS IoT broker
```
Same alignment issue as pointer above.

rlubos commented on include/net/aws_iot.h:
```diff
@@ -0,0 +1,185 @@
+ * @retval 0 If successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int aws_iot_init(struct aws_iot_config *const config,
```
`config` is documented as input parameter, should be `const struct aws_iot_config *`. Same with any other input pointer parameters in functions below.

rlubos commented on include/net/cloud.h:
```diff
@@ -260,6 +275,44 @@ static inline int cloud_input(const struct cloud_backend *const backend)
+/**@brief Add Cloud endpoint subscriptions.
+ *
+ * @param backend Pointer to cloud backend structure.
+ * @param endpoint Pointer to a list of endpoint subscriptions.
```
The description does not match the actual argument list.

simensrostad replied:
Fixed.

rlubos commented on include/net/cloud.h:
```diff
@@ -260,6 +275,44 @@ static inline int cloud_input(const struct cloud_backend *const backend)
+			struct cloud_endpoint *list, size_t list_count)
+{
+	if (backend == NULL
+	    || backend->api == NULL
```
Please keep the style consistent with other functions in this file.

simensrostad replied:
Fixed.

rlubos commented on include/net/cloud.h:
```diff
@@ -260,6 +275,44 @@ static inline int cloud_input(const struct cloud_backend *const backend)
+/**@brief Remove Cloud endpoint subscriptions.
+ *
+ * @param backend Pointer to cloud backend structure.
+ * @param endpoint Pointer to a list of endpoint subscriptions.
```
The description does not match the actual argument list.

simensrostad replied:
Fixed.

rlubos commented on include/net/aws_iot.h:
```diff
@@ -0,0 +1,185 @@
+int aws_iot_input(void);
+
+/** @brief Ping AWS IoT broker. Must be called periodically
+ *        to keep connection to broker alive.
```
Alignment

simensrostad replied:
Fixed.

rlubos commented on include/net/aws_iot.h:
```diff
@@ -0,0 +1,185 @@
+int aws_iot_ping(void);
+
+/** @brief Add a list of application spesific topics that will be subscribed to
+ *        upon connection to AWS IoT broker.
```
Alignment

simensrostad replied:
Fixed.

jtguggedal commented:
@rlubos No, I just forgot removing it after the offline discussions on how to proceed were finished.

ru-fu commented:
> 
> 
> Is this PR DNM still? @jtguggedal?

It needs a doc review before it can be merged. @umapraseeda will look at it.

jtguggedal commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,833 @@
+	broker4->sin_family = AF_INET;
+	broker4->sin_port = htons(CONFIG_AWS_IOT_PORT);
+
+	LOG_DBG("IPv4 Address %s", CONFIG_AWS_IOT_STATIC_IPV4_ADDR);
```
Missing `log_strdup`

jtguggedal commented on include/net/aws_iot.h:
```diff
@@ -0,0 +1,191 @@
+
+/** @brief AWS IoT library asynchronous event handler.
+ *
+ * @param[in] evt The event and any associated parameters.
```
`@param` is not aligned with `@brief`, same for the other functions in this file. Personally prefer to keep them aligned, but don't thing code style says anything about it though

simensrostad replied:
I agree, it looks better.

jtguggedal commented on include/net/cloud.h:
```diff
@@ -260,6 +275,42 @@ static inline int cloud_input(const struct cloud_backend *const backend)
+ */
+static inline int cloud_ep_subscriptions_add(
+			const struct cloud_backend *const backend,
+			struct cloud_endpoint *list, size_t list_count)
```
`const` list?

jtguggedal commented on include/net/cloud.h:
```diff
@@ -260,6 +275,42 @@ static inline int cloud_input(const struct cloud_backend *const backend)
+ */
+static inline int cloud_ep_subscriptions_remove(
+			const struct cloud_backend *const backend,
+			struct cloud_endpoint *list, size_t list_count)
```
`const` list?

simensrostad commented:
Documentation updates commited. @umapraseeda

jtguggedal commented:
@simensrostad Can you squash the doc commits into your own commit when they're done?

jtguggedal commented on subsys/net/lib/aws_iot/src/aws_iot.c:
```diff
@@ -0,0 +1,838 @@
+
+int aws_iot_ping(void)
+{
+	return mqtt_live(&client);
```
Let's use `mqtt_ping()` here instead, to avoid having to do workarounds to not go to much paste the kee-alive

simensrostad replied:
Agreed. updated
